### PR TITLE
feat(gfql): add case-insensitive support to startswith/endswith predicates (#697)

### LINKED ISSUE #697: [FEA] Add case-insensitive options and additional string predicates to GFQL
---
name: Feature request
about: Suggest an idea for this project
title: "[FEA] Add case-insensitive options and additional string predicates to GFQL"
labels: enhancement, gfql, help wanted, good-first-issue
assignees: ''

---

**Is your feature request related to a problem? Please describe.**

GFQL's string predicates have inconsistent support for case sensitivity and are missing some common string operations that users expect from a query language. Currently:

- `startswith()` and `endswith()` are always case-sensitive with no option to change this
- Several useful string predicates available in pandas are not exposed in GFQL
- The documentation doesn't clearly indicate which predicates support case-insensitive matching

**Describe the solution you'd like**

1. **Add case parameter to startswith() and endswith()**
   ```python
   # Current (case-sensitive only)
   n({"name": startswith("John")})
   
   # Proposed (with optional case parameter)
   n({"name": startswith("john", case=False)})  # Would match "John", "JOHN", etc.
   n({"email": endswith(".COM", case=False)})   # Would match ".com", ".Com", etc.
   ```

2. **Add commonly needed string predicates**
   ```python
   # Length checking
   n({"name": len_eq(5)})         # Exactly 5 characters
   n({"name": len_gt(10)})        # More than 10 characters
   n({"name": len_between(3, 20)}) # Between 3 and 20 characters
   
   # Strip/trim operations  
   n({"name": strip_eq("John")})   # Strip whitespace before comparing
   
   # Not/inverse operations
   n({"name": not_contains("test")})  # Doesn't contain pattern
   n({"name": not_match("^[0-9]+$")}) # Doesn't match regex
   ```

3. **Consider adding fuzzy/similarity matching** (stretch goal)
   ```python
   n({"name": similar_to("Jon", threshold=0.8)})  # Fuzzy string matching
   ```

**Describe alternatives you've considered**

1. **Query strings** - Users can use pandas query syntax, but this is less discoverable and not type-safe
2. **Custom predicates** - Users can implement their own, but built-in support would be better
3. **Post-processing** - Filter results after the query, but this is less efficient

**Additional context**

- The implementation should follow the existing pattern in `graphistry/compute/predicates/str.py`
- All new predicates should inherit from `ASTPredicate` and delegate to pandas string methods
- Tests should be added to ensure compatibility with both pandas and cuDF backends
- Documentation should be updated in `docs/source/gfql/spec/language.md`

**Implementation hints**

For case-insensitive startswith/endswith, the implementation could use pandas' case conversion:
```python
def startswith(self, pat, case=True, na=None):
    if not case:
        return self._series.str.lower().str.startswith(pat.lower(), na=na)
    return self._series.str.startswith(pat, na=na)
```

This would be a great first issue for someone familiar with Python and pandas who wants to contribute to GFQL!

### PR DESCRIPTION
**Pull request by lmeyerov**
## Summary

Adds `case` parameter and tuple pattern support to GFQL's `startswith()` and `endswith()` predicates, enabling case-insensitive string matching and OR logic in graph queries.

**Key Achievement**: ✅ **100% test pass rate (67/67 tests)** with full pandas/cuDF parity

Closes #697

### Case-Insensitive Matching
**Implementation** (graphistry/compute/predicates/str.py)
- Added `case: bool = True` parameter to `Startswith` and `Endswith` classes
- Used `str.lower()` workaround for case-insensitive matching (pandas/cuDF compatible)
- Fixed `na` parameter type from `Optional[str]` to `Optional[bool]` for consistency
- Added comprehensive docstrings with usage examples

**Example:**
```python
# Case-sensitive (default)
g.chain([n({"name": startswith("John")})])

# Case-insensitive
g.chain([n({"name": startswith("john", case=False)})])
```

### Tuple Pattern Support (OR Logic)
**Implementation** (graphistry/compute/predicates/str.py)
- Extended `pat` parameter to accept `str` or `tuple` of strings
- Tuple patterns provide OR logic: matches if string starts/ends with ANY pattern
- **cuDF bug workaround**: cuDF docs claim tuple support but implementation fails
  - Filed issue: https://github.com/rapidsai/cudf/issues/20237
  - Implemented manual OR logic for cuDF until bug fixed
  - pandas native tuple support used when available
- Full pandas/cuDF parity achieved with all edge cases handled

**Example:**
```python
# Single pattern
g.chain([n({"filename": startswith("test_")})])

# Multiple patterns (OR logic) - matches "test_*" OR "demo_*"
g.chain([n({"filename": startswith(("test_", "demo_"))})])

# Case-insensitive with multiple patterns
g.chain([n({"filename": startswith(("TEST", "DEMO"), case=False)})])
```

### Edge Cases Fixed
- Empty tuple `()` with `na=None` now preserves NA values correctly
- Object dtype used when NA values present to match pandas behavior
- cuDF bool dtype limitation handled with pandas intermediate conversion

### Tests
**Results**: ✅ **67/67 tests passing (100%)**
- 33/33 pandas tests PASSING
- 34/34 cuDF tests PASSING (GPU)
- All parity tests PASSING

**Coverage** (graphistry/tests/compute/predicates/test_str.py)
- Case-insensitive matching (both pandas and cuDF)
- Tuple pattern support (single and multiple patterns)
- Empty tuple edge cases
- NA handling with various `na` parameter values
- Case-insensitive + tuple pattern combinations
- Parity tests ensuring pandas/cuDF consistency

### Documentation
- Updated GFQL language specification grammar
- Updated quick reference with `case` parameter and tuple patterns
- Added inline examples for all usage patterns

### Changelog
- Added entries under `## [Development]` > `### Added`

## cuDF Bug Discovery

During implementation, discovered that cuDF documentation claims tuple pattern support but implementation fails:

**Documented**: `StringMethods.startswith(pat: str | tuple[str, ...]) → Series | Index`  
**Reality**: RuntimeError when tuple provided

- 🐛 **Filed issue**: https://github.com/rapidsai/cudf/issues/20237
- ✅ **Workaround**: Manual OR logic implemented in this PR
- 📋 **Tracking**: Will remove workaround when cuDF fixes the bug

## Test Results

```
GPU Tests (Docker):
  67 passed in 4.35s (100% pass rate) ✅
  - 33/33 pandas tests PASSED
  - 34/34 cuDF tests PASSED
  - All parity tests PASSED

CPU Tests (Local):
  33 passed, 34 skipped in 1.14s ✅
  - All pandas tests PASSED
  - cuDF tests skipped (expected, no GPU)
```

## Commits

- `cf65c31b` - Initial tuple pattern support
- `2e551d47` - cuDF tuple pattern workaround (97% pass rate)
- `e58d2c4b` - Empty tuple + NA parity fix (100% pass rate)
- `ed95f47d` - Add cuDF bug reference to code comments

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: ai/README.md [hunk 1]
```diff
@@ -165,10 +173,75 @@ WITH_BUILD=0 WITH_TEST=0 ./test-cpu-local.sh
...
+    $IMAGE \
+    ruff check graphistry/compute/predicates/str.py
+
+# Test cuDF-specific functionality with custom script
+docker run --rm --gpus all \
+    -v "$(pwd):/workspace:ro" \
+    -w /workspace \
+    -e PYTHONPATH=/workspace \
+    $IMAGE \
+    python3 plans/test_script.py
```
Comment (lmeyerov) [inline ai/README.md]: verbose
